aten/src/THC/generic/THCTensorIndex.cu

```diff
@@ -2,6 +2,62 @@
 #define THC_GENERIC_FILE "generic/THCTensorIndex.cu"
 #else
 
+// Check tensor dimensions for index operations, and return the slice size.
+// src can be nullptr in case of indexFill: in that case it is ignored.
+static ptrdiff_t THCTensor_(getSliceSize)(THCState *state, THCTensor *dst,
+                                          int dim,
+                                          THCudaLongTensor *index,
+                                          THCTensor *src)
+{
+  int dstDims = THCTensor_(nDimension)(state, dst);
+  int srcDims = (src == nullptr) ? dstDims : THCTensor_(nDimension)(state, src);
+
+  THArgCheck(THCudaLongTensor_nDimension(state, index) == 1, 4,
+             "expecting vector of indices");
+  THArgCheck(dim >= 0 && dim < dstDims, 2, "Indexing dim is out of bounds");
+
+  ptrdiff_t dstSliceSize = 1;
+  for (int d = 0; d < dstDims; d++) {
+    if (d != dim) {
+      dstSliceSize *= dst->size[d];
+    }
+  }
+
+  if (src == nullptr) return dstSliceSize;
+
+  THArgCheck(dim < srcDims, 3, "Indexing dim is out of bounds");
+  THArgCheck(THCudaLongTensor_nElement(state, index) == src->size[dim], 4,
+             "length of src.size[dim] is not equal to length of indices");
+
+  ptrdiff_t srcSliceSize = 1;
+  bool mismatch = false;
+
+  if (dstDims != srcDims) mismatch = true;
+
+  for (int d = 0; d < srcDims; d++) {
+    if (d != dim) {
+      srcSliceSize *= src->size[d];
+      if (!mismatch && dst->size[d] != src->size[d]) mismatch = true;
+    }
+  }
+
+  THArgCheck(dstSliceSize == srcSliceSize, 2,
+             "Source/destination tensor have different slice sizes (%ld vs %ld)",
+             dstSliceSize, srcSliceSize);
+
+  if (mismatch) {
```

Comment:
Ideally we would make THC strict as you proposed, and this check would live in the Python bindings (along with the code that reshapes the inputs correctly to fit into the strict constraints). Unconditionally printing from C is not a good practice, as the user has no control over it (in contrary to Python warnings)

Reply:
> (along with the code that reshapes the inputs correctly to fit into the strict constraints)

Hmm, I think it might be a bit difficult.  If the input or output is not contiguous, and the shapes do not match (say, (10, 5) vs (5, 10)), then the CUDA kernel *has* to know the dimensions of both input and output: the only way Python bindings could "reshape" input/output is by allocating a third buffer and copying data.

So I can think of several possible solutions:

(A) Make these operations as strict as possible: *any* shape mismatch triggers THArgCheck, even when slice sizes are the same.  May break backward compatibility.

(B) Migrate warning to Python bindings, which will merely print a warning and let the computation continue.  ATen still has to deal with the cases when shapes are different (as long as slice sizes are the same).

(C) Migrate warning to Python bindings: in case of shape mismatch, it will print a warning, allocate a buffer, and copy the result back to the original output.  ATen now THArgCheck's that the shapes match exactly.  (Sounds like too much work for borderline crazy cases, though.)

(D) Let's just remove `if (mismatch) { ... }` part here and add a TODO to deal with it in Python bindings.

Any opinions?

Reply:
@apaszke In case you missed it. :)